application-load-balancer/lambda_function/secret_rotator.py

```diff
@@ -0,0 +1,625 @@
+import json
+import os
+import boto3
+import logging
+import requests
+import time
+#import uuid
+from typing import Tuple, Dict, Any, List, Optional
+from slack_service import SlackNotificationService
+from requests.exceptions import RequestException
+
+from botocore.exceptions import ClientError
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+AWSPENDING = "AWSPENDING"
+AWSCURRENT = "AWSCURRENT"
+
+
+class SecretRotator:
+    def __init__(self, **kwargs):
+        # Use provided values or default to provided Lambda environment variables
+        self.secret_id = kwargs.get('secret_id', os.environ.get('SECRETID'))
+        self.waf_acl_name = kwargs.get('waf_acl_name', os.environ.get('WAFACLNAME'))
+        self.waf_acl_id = kwargs.get('waf_acl_id', os.environ.get('WAFACLID'))
+        waf_rule_priority = kwargs.get('waf_rule_priority', os.environ.get('WAFRULEPRI'))
+        try:
+            self.waf_rule_priority = int(waf_rule_priority)
+        except (TypeError, ValueError):
+            self.waf_rule_priority = 0
+        self.header_name = kwargs.get('header_name', os.environ.get('HEADERNAME'))
+        self.application = kwargs.get('application', os.environ.get('APPLICATION'))
+        self.environment = kwargs.get('environment', os.environ.get('ENVIRONMENT'))
+        self.role_arn = kwargs.get('role_arn', os.environ.get('ROLEARN'))
+        self.distro_list = kwargs.get('distro_list', os.environ.get('DISTROIDLIST'))
+        self.aws_account = kwargs.get('aws_account', os.environ.get('AWS_ACCOUNT'))
+
+        slack_token = kwargs.get('slack_token', os.environ.get('SLACK_TOKEN'))
+        slack_channel = kwargs.get('slack_channel', os.environ.get('SLACK_CHANNEL'))
+        self.slack_service = None
+        if slack_token and slack_channel:
+            self.slack_service = SlackNotificationService(slack_token, slack_channel,  self.aws_account)
+
+    def get_cloudfront_client(self) -> boto3.client:
```

Comment:
Do you actually want to call this function more than once?  I would expect that one assume role and client instantiation would be enough?  I notice that sometimes the client is being instantiated and then immediately instantiated again in a subsequent step.  You could consider moving (and in-lining) this function in the lambda handler, then passing in the client (which would also remove the need for the cloudfront role_arn dependency here.). It will mess up your tests though!  (you could also consider this a todo in a follow up refactor)

Reply:
This can be done in a follow up change.